### PR TITLE
[JENKINS-35098] Disable AutoBrowserHolder by default to improve the changelog rendering performance

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-35098](https://issues.jenkins-ci.org/browse/JENKINS-35098)

See JIRA for motivation.

Also took the opportunity to clean up some ancient form binding. The fewer weird tricks we have floating around the source base, the better.

@reviewbybees

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
I'm not convinced the AutoBrowser holder is useless at all.
My recommendation would be to...
1) deprecate the AutoBrowser class 
2) create AutoBrowser Holder only if SCM returns true for <code>isBrowserReusable()</code>
3) maybe it's possible to add SystemProperty to disable this AutoBrowser creation at all

**Comment by stephenc** [comment]
:bee: 👍 let's remove this "magic"

**Comment by jglick** [comment]
> create AutoBrowser Holder only if SCM returns true for `isBrowserReusable()`

Not so easy. This method takes both the “owner” SCM, and each candidate SCM. And the performance problem to begin with is finding those candidate SCMs. So you cannot actually call the method. The only thing you could do is use reflection to check if it is overridden.

Anyway this would only potentially benefit Subversion users (i.e., they would get the existing feature, at the existing performance cost). CVS users would continue to get the broken implementation, and the performance cost.

> add SystemProperty to disable

That would leave most users with the performance problem.

You could have a system property to _enable_ the feature, but by the time you figure out that you need this feature and such a property exists and you modify your Jenkins startup script and restart, you could have just fixed your job configurations already.

> I'm not convinced the AutoBrowser holder is useless

I did not say it was useless, just that on balance it seems like a poor decision from the start to have put this in core. We do not have anything else that I know of in Jenkins which quietly guesses what you meant to configure without you actually doing so. For people using traditional GUI configuration of jobs, this class of feature *can* be helpful, but the UI would probably need to be different: say, a list of suggested configuration tweaks to newly created jobs to set up not just the SCM browser but the `buildDiscarder`, mail notification, etc., with a way to batch-approve the suggestions or decline them. Fine territory for a plugin.

**Comment by oleg-nenashev** [comment]
@jglick 
OK, then SystemProperty, which enables this feature back.
I'm 👍 on disabling this feature by default.

**Comment by jglick** [comment]
Still prefer to delete it altogether and simplify the code, but if you insist…

File: core/src/main/java/hudson/scm/SCM.java [hunk 1]
```diff
@@ -85,6 +85,15 @@
 @ExportedBean
 public abstract class SCM implements Describable<SCM>, ExtensionPoint {
 
+    /** JENKINS-35098: discouraged */
+    @SuppressWarnings("FieldMayBeFinal")
+    private static boolean useAutoBrowserHolder = Boolean.getBoolean(SCM.class.getName() + ".useAutoBrowserHolder");
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/scm/SCM.java]: Please use SystemProperties.getBoolean() (new API in 2.4)

**Comment by oleg-nenashev** [comment]
🐝 from me, @reviewbybees done

File: core/src/main/java/hudson/scm/SCMS.java [hunk 2]
```diff
@@ -54,14 +54,11 @@
...
+    @SuppressWarnings("deprecation")
     public static SCM parseSCM(StaplerRequest req, AbstractProject target) throws FormException, ServletException {
-        String scm = req.getParameter("scm");
-        if(scm==null)   return new NullSCM();
-
-        int scmidx = Integer.parseInt(scm);
-        SCMDescriptor<?> d = SCM._for(target).get(scmidx);
-        d.generation++;
-        return d.newInstance(req, req.getSubmittedForm().getJSONObject("scm"));
+        SCM scm = req.bindJSON(SCM.class, req.getSubmittedForm().getJSONObject("scm"));
```
Comment (daniel-beck) [inline core/src/main/java/hudson/scm/SCMS.java]: @oleg-nenashev @jglick Could this have caused [JENKINS-35906](https://issues.jenkins-ci.org/browse/JENKINS-35906)?

Comment (jglick) [inline core/src/main/java/hudson/scm/SCMS.java]: Probably. No longer calling `SCMDescriptor.newInstance` overloads. They are a bad idea of course and unnecessary.

Comment (jglick) [inline core/src/main/java/hudson/scm/SCMS.java]: Possibly could use `DescriptorExtensionList.newInstanceFromRadioList` for this purpose.

**Pull request merged**